### PR TITLE
Modify timeout for etcd healthcheck

### LINKED ISSUE #111290: Allow configuration of etcd healthcheck in kube-apiserver
### What would you like to be added?

* Allow configuring separate etcd healthcheck timeouts for kube-apiserver's /readyz and /livez. 
* Allow disabling etcd healthcheck entirely in /livez.

### Why is this needed?

This is needed for more graceful degradation of kube-apiserver's performance when the control plane is overloaded.

Overloaded control plane can result in etcd slowness, and /readyz can be used to redirect traffic from it and allow it to recover. However, with failing /livez, kube-apiserver will soon be restarted by liveness probe anyway.

It might also be useful to stop liveness probe from restarting kube-apiserver at all when etcd is slow. Restarting kube-apiserver is unlikely to resolve problems with etcd, and can introduce extra disruption when kube-apiserver's performance is degraded.

### PR DESCRIPTION
**Pull request by Argh4k**
Increase default timeout for etcd healthcheck to 15 seconds.
Add additional etcd check to readyz with 2 seconds timeout.

#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
This PR increases default timeout for etcd healthcheck to 15 seconds and adds new etcd check to readiness check with timeout of 2 seconds. Currently, when the control plane is overloaded, healthchecks to etcd can take more than 2 seconds marking kube apsierver unhealthy, even if it is only etcd performance degradation. Adding 2 seconds check to readyz should help with load distribution between apiservers in case of etcd performance degradation.

#### Which issue(s) this PR fixes:
Fixes #111290

#### Does this PR introduce a user-facing change?

```release-note
a new flag `etcd-ready-timeout` has been added. It configures a timeout of an additional etcd check performed as part of readyz check. 
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/server/options/etcd.go [hunk 1]
```diff
@@ -234,6 +236,14 @@ func (s *EtcdOptions) addEtcdHealthEndpoint(c *server.Config) error {
 		return healthCheck()
 	}))
 
+	readyCheck, err := storagefactory.CreateReadyCheck(s.StorageConfig, c.DrainedNotify())
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:239]: I was wondering if we shouldn't use a different channel here, although given it's actually used as a stopCh it actually makes sense.

File: staging/src/k8s.io/apiserver/pkg/storage/storagebackend/config.go [hunk 2]
```diff
@@ -35,7 +35,8 @@ const (
 
 	DefaultCompactInterval      = 5 * time.Minute
 	DefaultDBMetricPollInterval = 30 * time.Second
-	DefaultHealthcheckTimeout   = 2 * time.Second
+	DefaultHealthcheckTimeout   = 15 * time.Second
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/storagebackend/config.go:38]: I understand the motivation and I'm supportive for the motivation as described in:
https://github.com/kubernetes/kubernetes/issues/111290

But technically that's a breaking change - for people who are not setting the timeout now, they will face a default behavior change.

For me it seems much safer to leave it set to 2s and just rely that people who would like to bump it will use the already existing flag. @deads2k - FYI

**Comment by wojtek-t** [comment]
I'm fine with this change modulo the comments that I added.

But I would like to give a bit of time to sig-apimachinery folks to take a quick look at it.

/sig api-machinery

File: staging/src/k8s.io/apiserver/pkg/server/options/etcd.go [hunk 3]
```diff
@@ -234,6 +236,14 @@ func (s *EtcdOptions) addEtcdHealthEndpoint(c *server.Config) error {
...
+	if err != nil {
+		return err
+	}
+	c.ReadyzChecks = append(c.ReadyzChecks, healthz.NamedCheck("etcd-readiness", func(r *http.Request) error {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:243]: we have a method for this

**Review by MadhavJivrajani — COMMENTED** [review]
Could we extend the existing tests to cover this change?: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/options/etcd_test.go

**Comment by sttts** [comment]
I can see why we want the readyz flag. I don't see why changing the healthz timeout from 2s to 15s and potentially break users who depend on the old behaviour.

File: staging/src/k8s.io/apiserver/pkg/storage/storagebackend/config.go (hunk 2 above)
Comment (mborsz) [reply staging/src/k8s.io/apiserver/pkg/storage/storagebackend/config.go:38]: I'm fine with leaving etcd as 2s and using existing flag to tune this.

File: staging/src/k8s.io/apiserver/pkg/server/options/etcd.go (hunk 3 above)
Comment (Argh4k) [reply staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:244]: I can only see AddHealthChecks which adds healthcheck to readyz/livez/healthz. It even has `We should prefer this to adding healthChecks directly to the config unless we explicitly want to add a healthcheck only to a specific health endpoint.` in its description. Am I missing something?

**Comment by Argh4k** [comment]
> I can see why we want the readyz flag. I don't see why changing the healthz timeout from 2s to 15s and potentially break users who depend on the old behaviour.
> 
> Also this change is not mentioned in the "Does this PR introduce a user-facing change?" section while it is userfacing.

Changed defaults back to original values. You are right that changing them, could break it for some users.

File: staging/src/k8s.io/apiserver/pkg/server/options/etcd.go (hunk 3 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/server/options/etcd.go:244]: Here is the function that @sttts was talking about:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/healthz.go#L63

But that is the function at the generic server level, whereas for etcd checks we're operating still at config level.

That said - instead of doing it manually, please create a AddReadyzCheck to the Config struct.

File: staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/factory_test.go [hunk 4]
```diff
@@ -0,0 +1,210 @@
+package factory
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/factory_test.go:1]: Boilerplate missing.

File: staging/src/k8s.io/apiserver/go.mod [hunk 5]
```diff
@@ -110,6 +110,7 @@ require (
 	golang.org/x/term v0.0.0-20210927222741-03fcf44c2211 // indirect
 	golang.org/x/text v0.3.7 // indirect
 	golang.org/x/time v0.0.0-20220210224613-90d013bbcef8 // indirect
+	golang.org/x/xerrors v0.0.0-20200804184101-5ec99f83aff1 // indirect
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/go.mod:113]: Why is this needed? You're not changing imports...

**Review by wojtek-t — COMMENTED** [review]
This looks good overall (modulo my one comment), but test failures seems related to this PR.

File: staging/src/k8s.io/apiserver/go.mod (hunk 5 above)
Comment (Argh4k) [reply staging/src/k8s.io/apiserver/go.mod:113]: I've run /hack/update-vendor.sh because `pull-kubernetes-dependecies` was failing with:
```
Your vendored results are different:
diff -Naupr -x BUILD -x 'AUTHORS*' -x 'CONTRIBUTORS*' vendor/k8s.io/apiserver/go.mod /home/prow/go/src/k8s.io/kubernetes/_tmp/kube-vendor.wBNJ6i/kubernetes/vendor/k8s.io/apiserver/go.mod
--- vendor/k8s.io/apiserver/go.mod	2022-07-27 07:20:52.922456943 +0000
+++ /home/prow/go/src/k8s.io/kubernetes/_tmp/kube-vendor.wBNJ6i/kubernetes/vendor/k8s.io/apiserver/go.mod	2022-07-27 07:23:22.144718498 +0000
@@ -110,6 +110,7 @@ require (
 	golang.org/x/term v0.0.0-20210927222741-03fcf44c2211 // indirect
 	golang.org/x/text v0.3.7 // indirect
 	golang.org/x/time v0.0.0-20220210224613-90d013bbcef8 // indirect
+	golang.org/x/xerrors v0.0.0-20200804184101-5ec99f83aff1 // indirect
 	google.golang.org/appengine v1.6.7 // indirect
 	google.golang.org/genproto v0.0.0-20220502173005-c8bf987b8c21 // indirect
 	google.golang.org/protobuf v1.28.0 // indirect
Vendor Verify failed.
If you're seeing this locally, run the below command to fix your directories:
hack/update-vendor.sh
```
I'm not quite sure why it is needed. I've changed formatting from %v to %w for error and used cmpopts.EquateErrors() but I do not understand why verify-vendor says that this should be included as indirect dependency.

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/go.mod:113]: Can you maybe revert that change then? I would like to avoid combining those two...

**Comment by wojtek-t** [comment]
/lgtm
/approve

Thanks!
